packages/andromeda-non-fungible-tokens/src/crowdfund.rs

```diff
@@ -1,111 +1,139 @@
 }
 
 #[andr_exec]
 #[cw_serde]
 pub enum ExecuteMsg {
-    /// Mints a new token to be sold in a future sale. Only possible when the sale is not ongoing.
-    Mint(Vec<CrowdfundMintMsg>),
-    /// Starts the sale if one is not already ongoing.
-    StartSale {
-        /// When the sale start. Defaults to current time.
-        start_time: Option<Expiry>,
-        /// When the sale ends.
-        end_time: Expiry,
-        /// The price per token.
-        price: Coin,
-        /// The minimum amount of tokens sold to go through with the sale.
-        min_tokens_sold: Uint128,
-        /// The amount of tokens a wallet can purchase, default is 1.
-        max_amount_per_wallet: Option<u32>,
-        /// The recipient of the funds if the sale met the minimum sold.
-        recipient: Recipient,
-    },
-    /// Updates the token address to a new one.
-    /// Only accessible by owner
-    UpdateTokenContract { address: AndrAddr },
-    /// Puchases tokens in an ongoing sale.
-    Purchase { number_of_tokens: Option<u32> },
-    /// Purchases the token with the given id.
-    PurchaseByTokenId { token_id: String },
-    /// Allow a user to claim their own refund if the minimum number of tokens are not sold.
-    ClaimRefund {},
-    /// Ends the ongoing sale by completing `limit` number of operations depending on if the minimum number
-    /// of tokens was sold.
-    EndSale { limit: Option<u32> },
+    /// Add a tier
+    AddTier { tier: Tier },
+    /// Update an existing tier
+    UpdateTier { tier: Tier },
+    /// Remove a tier
+    RemoveTier { level: Uint64 },
+
+    /// Start the campaign
+    StartCampaign {},
 }
 
 #[andr_query]
 #[cw_serde]
 #[derive(QueryResponses)]
-pub enum QueryMsg {
-    #[returns(State)]
-    State {},
-    #[returns(Config)]
-    Config {},
-    #[returns(Vec<String>)]
-    AvailableTokens {
-        start_after: Option<String>,
-        limit: Option<u32>,
-    },
-    #[returns(IsTokenAvailableResponse)]
-    IsTokenAvailable { id: String },
-}
+pub enum QueryMsg {}
 
 #[cw_serde]
-pub struct IsTokenAvailableResponse {
-    pub is_token_available: bool,
+pub struct CampaignConfig {
+    /// Title of the campaign. Maximum length is 64.
+    pub title: String,
+    /// Short description about the campaign.
+    pub description: String,
+    /// URL for the banner of the campaign
+    pub banner: String,
+    /// Official website of the campaign
+    pub url: String,
+    /// The address of the tier contract whose tokens are being distributed
+    pub tier_address: AndrAddr,
+    /// The denom of the token that is being accepted by the campaign
+    pub denom: String,
+    /// Withdrawal address for the funds gained by the campaign
+    pub withdrawal_address: AndrAddr,
+    /// The minimum amount of funding to be sold for the successful fundraising
+    pub soft_cap: Option<Uint128>,
+    /// The maximum amount of funding to be sold for the fundraising
+    pub hard_cap: Option<Uint128>,
```

Comment:
Publishing more tiers will lead to more rewards to tier owners. In case the campaign gains money more than enough and a good portion of them just remain untouched, it means the company need to pay rewards for the unnecessary funds that is just put aside. Or it can lead to a decrease in the value of the token.

Reply:
Is the hard cap not inherent in the tier structure?